### PR TITLE
Add MinimumPoint transpiler pass

### LINKED ISSUE #5832: Maximum iteration reached. max_iteration=1000
### Information

- **Qiskit Terra version**:

qiskit                        0.23.3             
qiskit-aer                    0.7.3              
qiskit-aqua                   0.8.1              
qiskit-ibmq-provider          0.12.0             
qiskit-ignis                  0.5.1              
qiskit-rng                    0.2.1              
qiskit-terra                  0.16.4 

- **Python version**:

python --version
Python 3.8.1

- **Operating system**:

cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

### What is the current behavior?

I am getting the same transpile optimization_level=3 error as here:
https://github.com/Qiskit/qiskit-terra/issues/5627
But I have updated to terra==0.16.4

More notes here:
https://qiskit.slack.com/archives/C7SS31917/p1612935056097500

### Steps to reproduce the problem

backend = provider.get_backend('ibmq_ourense')
noise_model = NoiseModel.from_backend(backend.properties())
qc = transpile(qc,optimization_level=3,\
                                   backend_properties=backend.properties(),\
                                   basis_gates=noise_model.basis_gates,\
                                   coupling_map=backend._configuration.coupling_map) 
                                   
it is a variational circuit, and if I only allow the parameters to evolve to the circ1 state (which has successfully transpiled,  circ1_untran(spiled) is before transpilation), the transpile optimization_level=3 succeeds (this is circ1).
If I allow the variational parameters to evolve further, eventually transpile will fail.

  File "/home/jmlarkin/.conda/envs/IQS/lib/python3.8/site-packages/qiskit/transpiler/runningpassmanager.py", line 311, in _iter_
    raise TranspilerError("Maximum iteration reached. max_iteration=%i" % self.max_iteration)
qiskit.transpiler.exceptions.TranspilerError: 'Maximum iteration reached. max_iteration=1000'

circ2_untran
circ1_untran
circ1

The circuit qasm definitions are here

[circ1.qasm.txt](https://github.com/Qiskit/qiskit-terra/files/5959970/circ1.qasm.txt)
[circ1_untran.qasm.txt](https://github.com/Qiskit/qiskit-terra/files/5959972/circ1_untran.qasm.txt)
[circ2_untran.qasm.txt](https://github.com/Qiskit/qiskit-terra/files/5959973/circ2_untran.qasm.txt)

### What is the expected behavior?

Expected behavior is ability to transpile the circuit with the above settings.  

### Suggested solutions

Not sure?

### LINKED ISSUE #9177: TranspilerError: 'Maximum iteration reached. max_iteration=1000'
### Environment

- **Qiskit Terra version**:  `'qiskit-terra': '0.22.2'`
- **Python version**:  `Python 3.10.6`
- **Operating system**:  `Windows 10`

### What is happening?

This error happens when specifying `seed_transpiler=0 ` during tranpilation of this Quantum Volume Circuit in Optimization level 3
Does not occur with optim level 2 or if adding an` initial_layout` 
Occurs on 27 qubit backend `Ehningen`, `mumbai`, `cairo`, `hanoi`, ...etc ... Does not occur for 7 qubit systems (`perth`, `mairobi`...`lagos `..etc...) 

### How can we reproduce the issue?

```
transpile(circuit,
          backend_properties=backend_properties,
          coupling_map=CouplingMap(coupling_map),
          optimization_level=3,
          basis_gates=basis_gates,
          seed_transpiler=0,
          # initial_layout=[0,1,2,3],
          callback=callback_func 
         )
```
Circuit qasm string : 
```
OPENQASM 2.0;
include "qelib1.inc";
qreg q[4];
creg measure[4];
rz(-2.5802401) q[0];
sx q[0];
rz(-2.0869703) q[0];
sx q[0];
rz(1.6404775) q[0];
rz(0.81085724) q[1];
sx q[1];
rz(-1.8376071) q[1];
sx q[1];
rz(-1.3097961) q[1];
cx q[1],q[0];
rz(0.67667501) q[0];
sx q[1];
rz(-2.7764038) q[1];
cx q[1],q[0];
rz(0.30962129) q[0];
sx q[1];
cx q[1],q[0];
rz(2.7770127) q[0];
sx q[0];
rz(-1.2031312) q[0];
sx q[0];
rz(-3.0338786) q[0];
rz(2.4776278) q[1];
sx q[1];
rz(-0.6905397) q[1];
sx q[1];
rz(2.0746922) q[1];
rz(-0.6156589) q[2];
sx q[2];
rz(-1.8610293) q[2];
sx q[2];
rz(-3.0344268) q[2];
rz(2.0113839) q[3];
sx q[3];
rz(-1.8378009) q[3];
sx q[3];
rz(-0.13322642) q[3];
cx q[3],q[2];
rz(1.016714) q[2];
sx q[3];
rz(-2.8928939) q[3];
cx q[3],q[2];
rz(0.60297329) q[2];
sx q[3];
cx q[3],q[2];
rz(-2.1374885) q[2];
sx q[2];
rz(-1.936627) q[2];
sx q[2];
rz(1.7758116) q[2];
cx q[2],q[0];
rz(-0.98539769) q[0];
sx q[2];
rz(-2.9594221) q[2];
cx q[2],q[0];
rz(0.17455528) q[0];
sx q[2];
cx q[2],q[0];
rz(-0.16450158) q[0];
sx q[0];
rz(-2.2954507) q[0];
sx q[0];
rz(-2.5644551) q[0];
rz(1.1609963) q[2];
sx q[2];
rz(-2.1767678) q[2];
sx q[2];
rz(2.5697281) q[2];
rz(-1.0296487) q[3];
sx q[3];
rz(-0.47555166) q[3];
sx q[3];
rz(1.9085924) q[3];
cx q[3],q[1];
rz(-0.61370581) q[1];
sx q[3];
rz(-2.8624277) q[3];
cx q[3],q[1];
rz(0.43911451) q[1];
sx q[3];
cx q[3],q[1];
rz(2.9516727) q[1];
sx q[1];
rz(-0.78588443) q[1];
sx q[1];
rz(-1.3890123) q[1];
cx q[1],q[0];
rz(-1.0649292) q[0];
sx q[1];
rz(-2.7754283) q[1];
cx q[1],q[0];
rz(0.46170102) q[0];
sx q[1];
cx q[1],q[0];
rz(-0.48364484) q[0];
sx q[0];
rz(-1.1779429) q[0];
sx q[0];
rz(1.8106294) q[0];
rz(-1.057089) q[1];
sx q[1];
rz(-0.48037255) q[1];
sx q[1];
rz(-2.3612646) q[1];
rz(-1.7485439) q[3];
sx q[3];
rz(-1.6016858) q[3];
sx q[3];
rz(-2.3028793) q[3];
cx q[3],q[2];
rz(1.093317) q[2];
sx q[3];
rz(-3.0564183) q[3];
cx q[3],q[2];
rz(0.11619561) q[2];
sx q[3];
cx q[3],q[2];
rz(0.17571848) q[2];
sx q[2];
rz(-1.735554) q[2];
sx q[2];
rz(-1.5351265) q[2];
rz(-2.5350841) q[3];
sx q[3];
rz(-0.81386815) q[3];
sx q[3];
rz(-3.0831125) q[3];
barrier q[0],q[1],q[2],q[3];
measure q[0] -> measure[0];
measure q[1] -> measure[1];
measure q[2] -> measure[2];
measure q[3] -> measure[3];
```

### What should happen?

Should transpile ! 

### Any suggestions?

_No response_

### LINKED ISSUE #9776: Max iteration being hit on QV circuits targetting ECR-based systems
### Environment

- **Qiskit Terra version**:  latest
- **Python version**: 3.9
- **Operating system**:  ubuntu

### What is happening?

It seems transpiling QV circuits for `ecr` based backends raises transpilation warnings for various seed values.  It is not hard to find a seed where this issue occurs.

### How can we reproduce the issue?

```python

from qiskit import *
from qiskit_ibm_provider import IBMProvider
from qiskit.circuit.library import QuantumVolume

provider = IBMProvider()

qc = QuantumVolume(7, seed=12345)
qc.measure_all()

backend = provider.get_backend('ibm_sherbrooke')
trans_qc = transpile(qc, backend, optimization_level=3)
```
gives

`TranspilerError: 'Maximum iteration reached. max_iteration=1000'`

Switching the backend to `ibm_kyiv` that has the same topology, but uses `cx` gates rather than  `ecr` works fine.

### What should happen?

It should work fine.

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit adds a new transpiler pass MinimumPoint which is used to find a local minimum point from the property set between executions of the pass. This is similar to the existing FixedPoint pass but will find the minimum fixed point over the past n exeuctions as opposed to finding a when two subsequent exectuions are at the same point. This is then used in optimization level 3 because the 2q unitary synthesis optimization that is part of the optimization loop can occasionally get stuck oscillating between multiple different equivalent decompositions which prevents the fixed point condition from ever being reached. By checking that we've reached the minimum over the last 5 executions we ensure we're reaching an exit condition in this situation.

### Details and comments

Fixes #5832
Fixes #9177
Fixes #9776

(the underlying cause of the optimization loop's inability to converge at optimization level 3 is not fixed here, there is still a root cause of instability in `UnitarySynthesis` this just changes the loop exit condition so we're never stuck in an infinte loop)

### CODE REVIEW

File: test/python/compiler/test_transpiler.py [hunk 1]
```diff
@@ -1362,7 +1362,7 @@ def test_transpile_optional_registers(self, optimization_level):
         out = transpile(qc, FakeBoeblingen(), optimization_level=optimization_level)
 
         self.assertEqual(len(out.qubits), FakeBoeblingen().configuration().num_qubits)
-        self.assertEqual(out.clbits, clbits)
+        self.assertEqual(len(out.clbits), len(clbits))
```
Comment (1ucian0) [inline test/python/compiler/test_transpiler.py:1365]: how come?

Comment (mtreinish) [reply test/python/compiler/test_transpiler.py:1365]: An equality check here is checking its the same object as the input. Prior to minimum point this would be true because the clbit is passed by reference through the dag to the output circuit. But with minimum point the output dagcircuit is deep copied meaning it's a different memory address/object than the original circuit's and this check would no longer be true.

File: qiskit/transpiler/passes/utils/minimum_point.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Check if the DAG has reached a fixed point."""
```
Comment (1ucian0) [inline qiskit/transpiler/passes/utils/minimum_point.py:13]:
```suggestion
"""Check if the DAG has reached a relative semi-stable point over previous runs."""
```

File: qiskit/transpiler/passes/utils/minimum_point.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        super().__init__()
+        self.property_set_list = property_set_list
+        self.backtrack_name = f"{prefix}_minimum_point_count"
+        self.backtrack_min_name = f"{prefix}_backtrack_history"
+        self.minimum_reached = f"{prefix}_minimum_point"
+        self.backtrack_depth = backtrack_depth
+        self.minimum_dag = None
+
+    def run(self, dag):
+        """Run the DAGFixedPoint pass on `dag`."""
```
Comment (1ucian0) [inline qiskit/transpiler/passes/utils/minimum_point.py:74]:
```suggestion
        """Run the MinimumPoint pass on `dag`."""
```

File: qiskit/transpiler/passes/utils/minimum_point.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    Fields used by this pass in the property set are (all relative to the ``prefix``
+    argument):
+
+     * ``{prefix}_minimum_point_count`` - Used to track the number of executions since
+        the current minimum was found
+     * ``{prefix}_backtrack_history`` - Stores the current minimum value and :class:`~.DAGCircuit`
+     * ``{prefix}_minimum_point`` - This value gets set to ``True`` when either a fixed point
+        is reached over the ``backtrack_depth`` executions, or ``backtrack_depth`` was exceeded
+        and an earlier minimum is restored.
+    """
```
Comment (1ucian0) [inline qiskit/transpiler/passes/utils/minimum_point.py:48]: This bullet list seems to not render correctly:

Screenshot 2023-03-11 at 15 08 21

File: test/python/transpiler/test_minimum_point.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""FixedPoint pass testing"""
```
Comment (1ucian0) [inline test/python/transpiler/test_minimum_point.py:13]:
```suggestion
"""MinimumPoint pass testing"""
```

File: qiskit/transpiler/passes/utils/minimum_point.py (hunk 4 above)
Comment (jakelishman) [reply qiskit/transpiler/passes/utils/minimum_point.py:48]: there's an extra space in the indentation, which is causing it to be parsed as a bullet point containing a single-item definition list

File: qiskit/transpiler/passes/utils/minimum_point.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            self.property_set[self.backtrack_name] += 1
+            if self.property_set[self.backtrack_name] == self.backtrack_depth:
+                self.property_set[self.minimum_reached] = True
+                return self.property_set[self.backtrack_min_name][1]
+        elif score < self.property_set[self.backtrack_min_name][0]:
+            self.property_set[self.backtrack_name] = 1
+            self.property_set[self.backtrack_min_name] = (score, deepcopy(dag))
+        elif score == self.property_set[self.backtrack_min_name][0]:
+            self.property_set[self.minimum_reached] = True
+            return self.property_set[self.backtrack_min_name][1]
```
Comment (jakelishman) [inline qiskit/transpiler/passes/utils/minimum_point.py:91]: Maybe it would be a shade cleaner to factor out all the different properties into something like a
```python
@dataclasses.dataclass
class _MinimumPointState:
    dag: DAGCircuit
    score: Tuple[float, ...]
    since: int
```
and store only that during the run?  I'm having trouble parsing all the different `self.property_set[self.backtrack_<x>]` stuff into the algorithm in my head.  The dataclass doesn't really solve that, it just removes a little bit of the boiler plate and gives things shorter names.

Then the loop of the pass would look something more like
```python
state = self.property_set[self.backtrack_state]
if state is None:
    state = self.property_set[self.backtrack_state] = _MinimumPointState(
        since=0,
        score=(math.inf,) * len(self.property_set_list),
        dag=None,
    )
if score < state.score:
    state.since = 1
    state.score = score
    state.dag = deepcopy(dag)
    return dag
if score > state.score:
    state.since += 1
if score == state.score or state.since == self.backtrack_depth:
    self.property_set[self.minimum_reached] = True
    return state.dag
return dag
```

I'm not certain I've got all the logic exactly correct there (because I find the current iteration a bit hard to parse), but certainly I think it's a shade easier to read.

File: test/python/compiler/test_transpiler.py (hunk 1 above)
Comment (jakelishman) [reply test/python/compiler/test_transpiler.py:1365]: Yeah, the fact that they were passed through perfectly before is more just a side effect of how little we touch `Clbit` instances in transpilation passes, rather than something we ought to be asserting as an invariant of transpile.

File: test/python/transpiler/test_minimum_point.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        """Test minimum resets backtrack depth."""
+        min_pass = MinimumPoint(["fidelity", "depth", "size"], prefix="test")
+        dag = DAGCircuit()
+        min_pass.property_set["fidelity"] = 0.875
+        min_pass.property_set["depth"] = 15
+        min_pass.property_set["size"] = 20
+        min_pass.run(dag)
+        self.assertEqual(min_pass.property_set["test_minimum_point_count"], 1)
+        self.assertIsNone(min_pass.property_set["test_backtrack_history"])
+        self.assertIsNone(min_pass.property_set["test_min_point"])
```
Comment (jakelishman) [inline test/python/transpiler/test_minimum_point.py:118]: Some of the assertions in these tests feel rather like testing relatively minor implementation details, like the exact objects we expect are tracked with very specific names in the property set.

Leaving that aside, though, the test is quite hard to quickly check what the sequence of values are being set as the properties; for something like this, I think it's maybe easier if we could define the sequence of inserted values and expected values up in single lists at the top of the test so all the context is localised, then loop through them.

**Review by jakelishman — COMMENTED** [review]
I'm happy with the logic of what this PR is trying to accomplish, and I'm happy that it'll be superior to what we've got in place already, although maybe we should replace _all_ use of `FixedPoint` with this - it feels like an accident waiting to happen in other places too.  My issues are mostly just with legibility - the logic is quite tricksy, so I think it's important that we put a good amount of effort into making both it and its tests as readable as possible.

File: test/python/transpiler/test_minimum_point.py (hunk 7 above)
Comment (mtreinish) [reply test/python/transpiler/test_minimum_point.py]: I left the tests mostly in place like this primarily to make it explicit what is happening in each iteration. I worry the details of the exact logic flow would be lost with a for loop. Especially given the complexity of the logic going on here my thinking was it might be difficult to track and testing it explicitly like this made it easier for me to ensure I got the expected logic. I agree that it does end up being a bit repetitive and a bit too detailed. In https://github.com/Qiskit/qiskit-terra/pull/9612/commits/e2c8cd023d86e30e48b32f28d46ba0db91b5d64d I added a bunch of comments to each iteration to explain the context.

File: qiskit/transpiler/passes/utils/minimum_point.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        # If the current execution is equal to the previous minimum value than
+        # we've reached an equivalent fixed point and we should use
+        elif score == state.score:
+            self.property_set[self.minimum_reached] = True
+            return dag
+
+        return dag
+
+
+@dataclass(slots=True)
```
Comment (jakelishman) [inline qiskit/transpiler/passes/utils/minimum_point.py:110]: Yeah, `slots` on `dataclass` is unfortunately Python 3.10+.

File: qiskit/transpiler/passes/utils/minimum_point.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        self.minimum_reached = f"{prefix}_minimum_point"
+        self.backtrack_depth = backtrack_depth
+
+    def run(self, dag):
+        """Run the MinimumPoint pass on `dag`."""
+        score = tuple(self.property_set[x] for x in self.property_set_list)
+        state = self.property_set[self.backtrack_name]
+
+        # The pass starts at None and the first iteration doesn't set a real
+        # score so the overall loop is treated as a do while to ensur we have
```
Comment (jakelishman) [inline qiskit/transpiler/passes/utils/minimum_point.py:81]:
```suggestion
        # score so the overall loop is treated as a do-while to ensure we have
```

File: qiskit/transpiler/passes/utils/minimum_point.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        # at least 2 iterations.
+        if state is None:
+            self.property_set[self.backtrack_name] = _MinimumPointState(
+                dag=None, score=(math.inf,) * len(self.property_set_list), since=0
+            )
+        # If the score of this execution is worse than the previous execution
+        # increment since since we have not found a new minimum point
```
Comment (jakelishman) [inline qiskit/transpiler/passes/utils/minimum_point.py:88]:
```suggestion
        # increment 'since' since we have not found a new minimum point
```

I think this might be what's meant?  It's hard to parse with the current punctuation.

File: qiskit/transpiler/passes/utils/minimum_point.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        elif score > state.score:
+            state.since += 1
+            if state.since == self.backtrack_depth:
+                self.property_set[self.minimum_reached] = True
+                return self.property_set[self.backtrack_name].dag
+        # If the score has decreased (gotten better) than this iteration is
+        # better performing and this iteration is a new local minimum. So
+        # update the state to be this iteration and reset counter
```
Comment (jakelishman) [inline qiskit/transpiler/passes/utils/minimum_point.py:96]:
```suggestion
        # If the score has decreased (gotten better) then this iteration is
        # better performing and this iteration is a new local minimum. So
        # update the state to be this iteration and reset counter
```
I suppose strictly we don't yet know that it's a "local minimum" so much as just "the best we've seen so far" (it could still be going downwards) but I'm not here to nitpick word choice in comments.

File: qiskit/transpiler/passes/utils/minimum_point.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        elif score < state.score:
+            state.since = 1
+            state.score = score
+            state.dag = deepcopy(dag)
+        # If the current execution is equal to the previous minimum value than
+        # we've reached an equivalent fixed point and we should use
```
Comment (jakelishman) [inline qiskit/transpiler/passes/utils/minimum_point.py:102]:
```suggestion
        # If the current execution is equal to the previous minimum value then
        # we've reached an equivalent fixed point and we should use
```

Comment (jakelishman) [reply qiskit/transpiler/passes/utils/minimum_point.py:102]: I think there's a word missing off the end here too, maybe?

File: test/python/transpiler/test_minimum_point.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+        self.assertIsNone(state.dag)
+        self.assertIsNone(min_pass.property_set["test_min_point"])
+
+        # Second iteration
+        min_pass.run(dag)
+        # After second iteration the state is initialized
+        state = min_pass.property_set["test_minimum_point_state"]
+        self.assertEqual(state.since, 1)
+        self.assertEqual(state.score, (42,))
+        self.assertIsNone(min_pass.property_set["test_min_point"])
```
Comment (jakelishman) [inline test/python/transpiler/test_minimum_point.py:45]: I think `test_min_point` should be `test_minimum_point` here? (And everywhere)

**Review by jakelishman — COMMENTED** [review]
The `slots` thing is Python 3.10+ only, but with that and the minor grammar tweaks in the comments for better legibility, this looks good to me.  The comments help a lot, and I think you're probably right about not wanting to do things in a loop in the tests.

File: qiskit/transpiler/passes/utils/minimum_point.py (hunk 8 above)
Comment (mtreinish) [reply qiskit/transpiler/passes/utils/minimum_point.py:110]: Sigh, that's too bad. I was going to add `__slots__` manually but then checked the docs and saw this argument. I'll remove this and just add the class attribute instead.

File: qiskit/transpiler/passes/utils/minimum_point.py (hunk 11 above)
Comment (mtreinish) [reply qiskit/transpiler/passes/utils/minimum_point.py:96]: Yeah, that's fair it has just gotten better, let me reword this:

```suggestion
        # If the score has decreased (gotten better) then this iteration is
        # better performing and this iteration should be the new minimum state.
        # So update the state to be this iteration and reset counter
```

File: qiskit/transpiler/passes/utils/minimum_point.py (hunk 12 above)
Comment (mtreinish) [reply qiskit/transpiler/passes/utils/minimum_point.py:102]:
```suggestion
        # If the current execution is equal to the previous minimum value then
        # we've reached an equivalent fixed point and we should use this iteration's
        # dag as the output and set the property set flag that we've found a minimum
        # point.
```

File: test/python/transpiler/test_minimum_point.py (hunk 13 above)
Comment (mtreinish) [reply test/python/transpiler/test_minimum_point.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/9612/commits/f889fbd11fcabe597688c34ba756bdc6d69f582b

File: qiskit/transpiler/passes/utils/minimum_point.py (hunk 8 above)
Comment (mtreinish) [reply qiskit/transpiler/passes/utils/minimum_point.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/9612/commits/a339bb1964e467da3ab711839d9049f5dde57b7a

**Review by jakelishman — APPROVED** [review]
Ace, thanks!
